bin/varnishd/cache/cache_obj.c

```diff
@@ -220,17 +222,33 @@ void
 ObjExtend(struct worker *wrk, struct objcore *oc, ssize_t l, int final)
 {
 	const struct obj_methods *om = obj_getmethods(oc);
+	AN(om->objextend);
 
 	CHECK_OBJ_NOTNULL(wrk, WORKER_MAGIC);
 	CHECK_OBJ_NOTNULL(oc->boc, BOC_MAGIC);
 	(void)final;
 	assert(l >= 0);
 
 	Lck_Lock(&oc->boc->mtx);
-	AN(om->objextend);
 	if (l > 0) {
+		/* Pass prefetch applies when only sending to one client */
+		if (oc->flags & OC_F_PRIVATE) {
+			/* delivered_so_far will be re-read due to memory clobber
+			   from the CondWait opaque function call */
+			while (!(oc->flags & OC_F_ABANDON) &&
+				oc->boc->pass_prefetch > 0 &&
+				oc->boc->fetched_so_far > oc->boc->delivered_so_far + oc->boc->pass_prefetch) {
+				/* Wait 2.500s for client to catch up */
```

Comment:
I think this comment isn't clear. Also the timeout is on the high side I think, maybe 0.1s would be better? (This controls for how long a stuck fetch thread could be sitting waiting for a gone client).

Suggested new comment wording:

The client is too far behind, wait for it to catch up and signal the condvar. Lck_CondWait() with timeout is used to catch any failure to signal the condvar on the client's part (e.g. a restart or late hangup detection), avoiding deadlocks.